Menu Plugins For external sites

martinpovolny opened the pull request:
## Using Menu Plugins ##

For plugins that implement integration with an external web site we need a possibility to define:

 1. Menu sections and menu items
  1. for top_right menu
  1. for main menu
 1. RBAC features for the menu items that are editable using Config --> Configuration --> Role

Custom menu sections and items can be defined under `vmdb/product/menubar`.

Custom RBAC features can be defined under `vmdb/db/fixtures/miq_product_features/`.

Assignments of the custom RBAC features to pre-defined roles can be defined under `vmdb/db/fixtures/miq_user_roles`.

For the current version of ManageIQ a package with a UI plugin should drop appropriate files into these locations.

**A custom menu section definition is a yaml file with these keys:**
```
type: section
name: Red Hat
id: redhat
before: svc
placement: top_right
section_type: big_iframe
href: http://www.redhat.com
```

 * `type`, `name`, `id` are mandatory
 * `placement` with value of `top_right` means top-right menu, omitted or `default` means main menu
 * `section_type` `default` creates a menu section with the same type as built-in menu sections; `big_iframe` requires a defined `href` and creates a menu section that when clicked replaces the 2nd menu level and all below with an iframe containing site given in the `href`. This is only valid for main menu items.
 * `before` allows to define the position of the menu section relative to other menu secrion

**A custom menu item definition is a yaml file withe these keys:**
```
type: item
name: Homepage
id: rht3
feature: redhat
rbac:
  feature: redhat
parent: redhat
href: http://www.redhat.com
item_type: big_iframe
```
 * `type`, `name`, `id`, `parent`, `feature` and `rbac` are mandatory
 * `feature` has to be defined as RBAC feature
 * `rbac` contains arguments passwd to `role_allows?` call when a decision about whether a menu item should be presented to a particular user is done
 * `item_type` and `href` ha similar meaning as `section_type` above
 * `parent` must reference an existing menu section


---

DONE:
  * Menu Section definition including `default` and `top_right` *placement*. And possibility to specify ordering by *after*.
  * Settings/Access Control/Role knows about items that exist as features and can edit them.
  * Allow adding of RBAC features (now only existing features will be saved when editing a Role in Settings/Access Control/Role which is enough for testing but not for production)
  * Styling

Future work:
 * restore CSP header and X-Frame-Options header; now these are disabled for `/dashboard/iframe`, we should just relax them by adding some URLs for the plugin
 * maybe? Implement `rake` task for post-inst scripts for plugins

Example usage
----

**product/menubar/custom_redhat_section.yml**
```
type: section
name: Red Hat
id: redhat
before: svc
section_type: big_iframe
href: http://www.redhat.com
```

**product/menubar/custom_redhat.yml**
```
type: item
name: Homepage
id: rht3
feature: redhat
rbac:
  feature: redhat
parent: redhat
href: http://www.redhat.com
item_type: big_iframe
```

**product/menubar/custom_redhat_courses.yml**
```
type: item
name: Courses
id: rht2
feature: redhat
rbac:
  feature: redhat
parent: redhat
href: http://www.redhat.com/en/services/training/all-courses-exams
item_type: big_iframe
```

![custom-menu](https://cloud.githubusercontent.com/assets/51095/6288736/64631086-b91b-11e4-8a56-7cc437f52807.png)

 1. custom menu based on the definition above
 1. settings in the `Configure/Configuration/Access Control/Roles` (the "About" is replaced with the correct text in current PR)
 1. custom menu with *placement*: `top_right`                                                                                

**product/menubar/top_right_section1.yml**
```
type: section                                                                                       
name: Custom section 1                                                                              
placement: top_right                                                                                
id: custom_top  
```

**db/fixtures/miq_product_features/redhat-miq_product_features.yml**
```
---
:name: Red Hat
:description: Access to Red Hat sectioin
:feature_type: node
:identifier: redhat
:children:

- :name: Red Hat Forums
  :description: Access to Red Hat forums
  :feature_type: node
  :identifier: redhat_forum
- :name: Red Hat Whatever
  :description: Access to Red Hat whatever...
  :feature_type: node
  :identifier: redhat_whatever
```

**db/fixtures/miq_user_roles/redhat.yml**
```
--- 

- :name: EvmRole-super_administrator
  :miq_product_feature_identifiers:
  - redhat
  - redhat_forum
  - redhat_whatever
```

![menu-new](https://cloud.githubusercontent.com/assets/51095/6485935/98a477fc-c287-11e4-8c5c-2aa66e03acb6.png)

martinpovolny commented:
@Fryguy , @dclarizio : Please, see 2 commits dealing with fixture.

I need to be able to add features and feature-role relations when installing a plugin. I cannot simply add the values to the tables such as `miq_product_features`. This would not work in development mode because some tables (including `miq_product_features`) are seeded each time the web worker is started.
Fixtures are loaded from `db/fixtures/<model_name>.uml`
Furthermore the seeding of features is done is such a way that features in the DB but not in the fixture file are deleted in the process.

So I extended the seeding for some 2 models by including all files under `db/fixtures/<model_name>/`.

So a pluging (e.g. packaged as an RPM) might 
  1. simply drop the extra fixtures into the corresponding fixture directories. 
  2. call some `rake` task (to be implemented) to load the fixtures also into the production database -- this would be done in a RPM post-inst script.

I believe the above satisfies the basic needs of the UI plugin with an iframe. I welcome any feedback.

Fryguy commented:
First, the term "model_name" is confusing because when I think models I think AR models.  So, I'll refer to them as "plugin_name" below to prevent the confusion.

I think the seeding should change to allow "system" records vs "user" records.  We've done this for many of the other tables by introducing a "system" boolean column, and then restricting the seeding to only the records `where(:system => true)`.  For user records, you may also want to qualify them by a plugin name or something

Having the external product features bucketed seems good, but I'm not sure bucketing under `db/fixtures/<plugin_name>/` is a good idea.  I'm concerned because then custom values could collide with ManageIQ things since they are in the same directory structure.  For example we already have a db/fixtures/ae_datastore, so won't that confuse the code?  It should probably live external, like in a gem or Rails engine, with some way for the code to "register" itself.  Then from that registered location we can load all of the custom stuff.

Additionally in a git enabled appliance, the extra directory would appear as an addition in git, making using the appliance as a dev more difficult.

Also, can the "seed on startup" still be leveraged?  Then there would be no need for the RPM post-init thing.  The seed on startup would just go to all registered directories and seed the fixtures.

martinpovolny commented:
@Fryguy : Paragraph 1: No, I am talking about "model_name" as in AR model. I need to load `miq_product_features` and `miq_user_role` fixtures.

2: From the point of view of the user we are talking only about "system" records. Some product will be integrated into ManageIQ and from the point of view of the user ManageIQ + the plugin are the same, that's the point.

Paragraph 3: is partly not relevant due to a misunderstanding in paragraph 1. Please, see the examples of the content of the .yml files in description above. Partly yes, as a developer you would see those files as untracked files under vmdb/product. 

And yes, we may need some plugin architecture maybe based on Rails engines. But I believe this is a larger task for some future work.

Last paragraph: Well yes, the startup mechanism would work if you install the plugin RPM before first running the appliance I believe. In that case the additional fixtures would be there and we would not need to have any rake task to load the data.

But I am thinking about a scenario where the plugin RPM/package wold be installed later when the appliance is already being uses. Maybe I can skip that scenario for now? If there would be an appliance with CFME and the plugin installed that would be delivered to the customer... What do you think?

But it would be just convenient to have a way to load the additional fixtures even when you are a developer. The other table -- `miq_user_role` does not do the automatic seeding on restart of `rails s` if I am reading it right.

dclarizio commented:
@martinpovolny @Fryguy 
On the topic of seeding the product features:
* It is fine if the plugin features seed at the same time the included features seed, as installing a plugin is not a dynamic thing.  In other words, I would EXPECT an appliance restart being needed to install a plugin, which should reseed all tables.
* Plugin features should act the same as included features in that if the plugin is removed, the plugin's features should be removed from the table during seeding.
* In development mode, the developer can manually run any model's seed method, if adding/changing the plugin's seed file(s).

Fryguy commented:
> No, I am talking about "model_name" as in AR model. I need to load miq_product_features and miq_user_role fixtures.

Ohhh I see.  So you could add multiple files into db/fixtures/miq_product_features like db/fixtures/miq_product_features/plugin1.yml  That's pretty nice.

> 2: From the point of view of the user we are talking only about "system" records. Some product will be integrated into ManageIQ and from the point of view of the user ManageIQ + the plugin are the same, that's the point.

Then maybe instead of a boolean `system` column, maybe a `scope` string column where "system" means *our* records, and "plugin1" means it came from that plugin1.yaml.  Then the seeding is scoped, and the same add/update/delete logic can still apply without stepping on each other.

>  But I believe this is a larger task for some future work

Agreed, but we should at least think about that future stuff and not paint ourselves into a corner...I'm concerned that the external content living in our directories will make upgrades more difficult or could potentially get in the way of any content *we* may need to drop into those directories.

Is it too much of a stretch to have the files live external, and then have some way to "register" plugin locations?  It's definitely work, but it doesn't sound like a lot...maybe?  It could be a separate PR, too, so we can think about it.

>  If there would be an appliance with CFME and the plugin installed that would be delivered to the customer... What do you think?

Yes, I was thinking that scenario.  Even so, if it was added later, a simple restart of EVM (or even just the UI workers) would go through the seeding logic.

dclarizio commented:
@martinpovolny OK, I "almost" got this working on the top right I think:

**product/menubar/top_right_news.yml**
```
type: section
name: News
placement: top_right
id: news_top
feature: news
rbac:
  feature: news
```
**product/menubar/top_right_news_cnn.yml**
```
type: item
name: CNN
placement: top_right
id: news_cnn
feature: news_cnn
rbac:
  feature: news_cnn
parent: news_top
href: http://www.cnn.com
item_type: big_iframe
```
**db/fixtures/miq_product_features/top_right_news_features.yml**
```
---
:name: News
:description: Access to News
:feature_type: node
:parent: everything
:identifier: news
:children:
- :name: CNN
  :description: CNN News Site
  :feature_type: node
  :identifier: news_cnn
```
Results in this menu showing:
![top right menu](https://cloud.githubusercontent.com/assets/3433754/6495527/f1d80314-c27f-11e4-93d4-d32868fbdfc0.png)
So far, so good, but clicking on News/CNN takes me to the page, but not in an iFrame:
![cnn](https://cloud.githubusercontent.com/assets/3433754/6495547/237995fe-c280-11e4-9a2f-0934b5cd9bb3.png)

So:
* What am I configuring wrong?
* Is there any way to just make News an item and not a section in the top right? (not saying do it, I'm not sure if that's a requirement or not)

dclarizio commented:
@martinpovolny Tried on the menu bar and success when clicking News -> CNN:

![cnn_iframe](https://cloud.githubusercontent.com/assets/3433754/6495919/05c107c4-c283-11e4-86f5-fa7742055239.png)

However, clicking on News main tab gets (so some bug here):

![news_menu_click](https://cloud.githubusercontent.com/assets/3433754/6495951/486294ee-c283-11e4-8a42-d0bc76ac6b34.png)

martinpovolny commented:
@dclarizio : that's the site. The site "frees" itself from the iframe. Many big sites so that.

martinpovolny commented:
@dclarizio : I will try your files and see

martinpovolny commented:
@dclarizio : I have fixed the URL for the top_right section menu items. Thx for finding the bug.

I have changed the logic of section placement from `after` to `before`. Now you have to specify `before` or the section will be placed as the last one.

a couple of notes
 1. The `placement: top_right` belongs only to `section`
 1. The top_right sections ARE NOT clickable while the main menu sections ARE clickable. Therefor for main menu section you have 2 options
  1. you specify `section_type: big_iframe` and `url: .....` and you get the same behavior as for items or
  2. you don't -- you get the behavior or the "system" menu sections -- redirect to `/dashboard/maintab/?tab=<ID>`; this makes little sense now but will make more sense in the future when we support plugins with own controllers etc.

dclarizio commented:
@martinpovolny Tested out and the main tab looks pretty good.  I did have to add:
```
section_type: big_iframe
href: http://www.cnn.com
```
to my section yaml file to get it to not show the login screen when clicking on the main tab.
If I added only the section_type, I get a blank main area when clicking on the main tab.

I'm thinking a follow up PR such that if there is no section_type or href specified, the main tab becomes just text, not a link.  Alternatively, we could remember the last subtab chosen and direct the user to either the first tab (if none already visited) or the last one visited.

dclarizio commented:
@martinpovolny The top right tab is not being enforced by RBAC, I think we need this working before I merge the PR.

martinpovolny commented:
@dclarizio :fixed the RBAC on top rigtht; thx for the review!

dclarizio commented:
Everything looks good!
